agent/src/agent/pipeline/config/jython_scripts/zabbix.py

```diff
@@ -74,57 +79,140 @@ def get_interval():
     return int(sdc.userParams['INTERVAL'])
 
 
+def get_last_processed_id():
+    if sdc.lastOffsets.containsKey(entityName):
+        return sdc.lastOffsets.get(entityName).split('_')[1]
+
+
+def query_items_sorted():
+    # returns [{itemid: val, value_type: val}, ...]
+    data = client.post('item.get', json.loads(sdc.userParams['QUERY']))
+    if len(data) == 0:
+        sdc.log.info('item.get - No data - query: ' + str(json.loads(sdc.userParams['QUERY'])))
+    return data
+
+
+def query_history(item_ids, value_type):
+    for ids_chunk in chunks(item_ids, HISTORIES_BATCH_SIZE):
+        sdc.log.info('fetching history for ' + str(len(ids_chunk)) + ' items')
+        history_params = {
+            'history': value_type,
+            'itemids': ids_chunk,
+            'sortfield': 'clock',
+            'sortorder': 'ASC',
+            'time_from': end - interval,
+            'time_till': end
+        }
+        histories = client.post('history.get', history_params)
+        sdc.log.info('fetched history')
+        # add fields from item to every history record
+        if len(histories) == 0:
+            sdc.log.info('history.get - No data - query: ' + str(history_params))
+            continue
+        yield histories
+
+
+def chunks(lst, n):
+    for i in range(0, len(lst), n):
+        yield lst[i:i + n]
+
+
+def fetch_hosts():
+    res = client.post('host.get', {'output': ['hostid', 'name']})
+    if len(res) == 0:
+        sdc.log.info('host.get - No data')
+    hosts = {}
+    for host in res:
+        hosts[host['hostid']] = host
+    return hosts
```

Comment:
`return {host['hostid']: host for host in res}`

this is not necessary, but just in case you didn't know :)

Reply:
yep, thanks, I was too lazy to search for this solution)